packages/javascript/bh-shared-ui/src/components/CreateUserForm/CreateUserForm.tsx

```diff
@@ -14,426 +14,699 @@
+}> = ({ error, isLoading, onSubmit, showEnvironmentAccessControls }) => {
+    const defaultValues = {
+        email_address: '',
+        principal: '',
+        first_name: '',
+        last_name: '',
+        password: '',
+        needs_password_reset: false,
+        roles: [3],
+        sso_provider_id: '',
+        all_environments: false,
+        environment_targeted_access_control: {
+            environments: null,
         },
-    });
+    };
+
+    const form = useForm<CreateUserRequestForm>({ defaultValues });
 
     const [authenticationMethod, setAuthenticationMethod] = React.useState<string>('password');
+    const [needsPasswordReset, setNeedsPasswordReset] = useState(false);
+    const [searchInput, setSearchInput] = useState<string>('');
+    const [selectedEnvironments, setSelectedEnvironments] = useState<string[]>([]);
+    const [selectedRoleValue, setSelectedRoleValue] = useState([3]);
 
-    useEffect(() => {
-        if (authenticationMethod === 'password') {
-            setValue('SSOProviderId', undefined);
+    const handleCheckedChange = (checked: boolean | 'indeterminate') => {
+        setNeedsPasswordReset(checked === true);
+        if (checked === true) {
+            form.setValue('needs_password_reset', true);
+        } else {
+            form.setValue('needs_password_reset', false);
+        }
+    };
+
+    const getRolesQuery = useQuery(['getRoles'], ({ signal }) =>
+        apiClient.getRoles({ signal }).then((res) => res.data?.data?.roles)
+    );
+
+    const listSSOProvidersQuery = useQuery(['listSSOProviders'], ({ signal }) =>
+        apiClient.listSSOProviders({ signal }).then((res) => res.data?.data)
+    );
+
+    const matchingRoles = getRolesQuery.data
+        ?.filter((item) => selectedRoleValue.includes(item.id))
+        .map((item) => item.name);
+
+    const selectedETACEnabledRole = matchingRoles?.toString() === 'Read-Only' || matchingRoles?.toString() === 'User';
+
+    const { data: availableEnvironments } = useAvailableEnvironments();
+
+    const filteredEnvironments = availableEnvironments?.filter((environment: Environment) =>
+        environment.name.toLowerCase().includes(searchInput.toLowerCase())
+    );
+
+    const allEnvironmentsSelected: any =
+        selectedEnvironments &&
+        selectedEnvironments.length === availableEnvironments?.length &&
+        availableEnvironments!.length > 0;
+
+    const handleSelectAllEnvironmentsChange = (allEnvironmentsChecked: string | boolean) => {
+        if (allEnvironmentsChecked || allEnvironmentsIndeterminate) {
+            const returnMappedEnvironments: string[] | undefined = availableEnvironments?.map(
+                (environment) => environment.id
+            );
+            returnMappedEnvironments && setSelectedEnvironments(returnMappedEnvironments);
+        } else {
+            setSelectedEnvironments([]);
         }
+    };
 
+    const formatReturnedEnvironments: EnvironmentRequest[] =
```

Comment:
looks like this variable is only used in the handleOnSave function? If so, could we move the definition into that function so we dont redefine the variable on every rerender?